Add new version for all AOCL libs and update aocl_libmem and amdlibm

amd-toolchain-support opened the pull request:
This PR includes below main changes in addition to few minor supporting changes:
- Adding new version 5.1 for AOCL libraries: 
`amdblis` `amdfftw` `amdlibflame` `amdlibm` `amdscalapack` `aocl-compression` `aocl-crypto` `aocl-da` `aocl-libmem` `aocl-sparse` `aocl-utils` and corresponding bundle package `amd-aocl`
- Adding `CMake` support for `amdlibm` with version `5.1`
- Update compiler dependency for `amdlibm`
- Update `aocl-libmem` recipe to include new variants `dynamic-dispatch` and update existing variant `vectorization` specific to version `5.1`
- Align the CMake version dependency for all libraries.

amd-toolchain-support committed:
Add new version for all AOCL libs and update aocl_libmem and amdlibm

amd-toolchain-support committed:
Style fix

amd-toolchain-support commented:
@adamjstewart, @wdconinc adding you as reviewer based on suggestion.

adamjstewart commented:
I don't know anything about these packages, will let someone else review.

wdconinc commented:
I'm going to have to express the same reservations as Adam; I don't even have amd systems to test on...

I'd suggest posting under #pull-requests on slack.

wdconinc commented on repos/spack_repo/builtin/packages/amd_aocl/package.py:
```diff
@@ -25,7 +25,8 @@ class AmdAocl(BundlePackage):
     maintainers("amd-toolchain-support")
 
-    version("5.0", preferred=True)
+    version("5.1", preferred=True)
```
Preferred doesn't do anything when it's on the most recent version.
```suggestion
    version("5.1")
```

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdblis/package.py:
```diff
@@ -37,10 +37,11 @@ class Amdblis(BlisBase):
 
     version(
-        "5.0",
-        sha256="5abb34972b88b2839709d0af8785662bc651c7806ccfa41d386d93c900169bc2",
+        "5.1",
+        sha256="4ab210cea8753f4be9646a3ad8e6b42c7d19380084a66312497c97278b8c76a4",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.
```suggestion
    version("5.1", sha256="4ab210cea8753f4be9646a3ad8e6b42c7d19380084a66312497c97278b8c76a4")
```

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdfftw/package.py:
```diff
@@ -40,10 +40,11 @@ class Amdfftw(FftwBase):
 
     version(
-        "5.0",
-        sha256="bead6c08309a206f8a6258971272affcca07f11eb57b5ecd8496e2e7e3ead877",
+        "5.1",
+        sha256="4cc0d6985afc5ce14cafc195ad46d9876ac4f35b959861b41a3b681385d7320d",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdlibflame/package.py:
```diff
@@ -49,10 +49,11 @@ class Amdlibflame(CMakePackage, LibflameBase):
 
     version(
-        "5.0",
-        sha256="3bee3712459a8c5bd728a521d8a4c8f46735730bf35d48c878d2fc45fc000918",
+        "5.1",
+        sha256="25524ba78b5952303369fa0859d217e44071144fd122a9dc3f72ed0bd73e3b2d",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdlibflame/package.py:
```diff
@@ -83,7 +84,7 @@ class Amdlibflame(CMakePackage, LibflameBase):
     with when("build_system=cmake"):
         generator("make")
-        depends_on("cmake@3.22:", type="build")
+        depends_on("cmake@3.26:", type="build")
```
This needs to be added with a `when` clause, and the previous cmake minimum version should remain for older version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdlibm/package.py:
```diff
@@ -33,10 +34,11 @@ class Amdlibm(SConsPackage):
 
     version(
-        "5.0",
-        sha256="ba1d50c068938c9a927e37e5630f683b6149d7d5a95efffeb76e7c9a8bcb2b5e",
+        "5.1",
+        sha256="7acf2c98469353b60a59fad167a98e1ae689055a3faf8352a254832145c9d59e",
         preferred=True,
     )
```
(You guessed it) Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdlibm/package.py:
```diff
@@ -47,14 +49,17 @@ class Amdlibm(SConsPackage):
     variant("verbose", default=False, description="Building with verbosity", when="@:4.1")
 
+    # Build system
+    build_system(conditional("cmake", when="@5.1"), "scons", default="scons")
```
Unless you'll remove this for the next version:
```suggestion
    build_system(conditional("cmake", when="@5.1:"), "scons", default="scons")
```

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/amdlibm/package.py:
```diff
@@ -63,12 +68,10 @@ class Amdlibm(SConsPackage):
-    conflicts("%clang@:9.0", msg="Minimum supported Clang version is 9")
-    conflicts("%clang@17.0.0:", msg="Maximum supported Clang version is 17.0.0")
-    conflicts("%gcc@14.3.0:", msg="Maximum supported GCC version is 14.2.0")
+    conflicts("%clang@:11.1", msg="Minimum supported Clang version is 12.0")
```
Previous conflicts should be retained but guarded with `when="@:5.0"`.

amd-toolchain-support replied:
compiler version dependency for `amdlibm` is updated for:
- Removing the conflicts for the higher version. Earlier it was used to align with list of supported compilers from user guide. However, there is no specific reason why this should not work with recent compiler versions. We do not want the block end users to try more recent compilers and report issues.
- Change in the minimum clang compiler version in a correction in previous conflicts which was wrong according to the documentation.

wdconinc commented on repos/spack_repo/builtin/packages/amdscalapack/package.py:
```diff
@@ -33,11 +33,13 @@ class Amdscalapack(ScalapackBase):
+
     version(
-        "5.0",
-        sha256="a33cf16c51cfd65c7acb5fbdb8884a5c147cdefea73931b07863c56d54f812cc",
+        "5.1",
+        sha256="92f6f6b2081e27731c8b9e96c742203777cc7f2a848b96ca7511f4d259142b37",
         preferred=True,
     )
```
Here too. Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_compression/package.py:
```diff
@@ -45,10 +45,11 @@ class AoclCompression(CMakePackage):
 
     version(
-        "5.0",
-        sha256="50bfb2c4a4738b96ed6d45627062b17bb9d0e1787c7d83ead2841da520327fa4",
+        "5.1",
+        sha256="9462c6898350d66a5d9ce0236c432c94b4c9393b638ddf6511628b784eb02720",
         preferred=True,
     )
```
And here: Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_compression/package.py:
```diff
@@ -76,7 +77,7 @@ class AoclCompression(CMakePackage):
     depends_on("cxx", type="build")  # generated
 
-    depends_on("cmake@3.22:", type="build")
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_crypto/package.py:
```diff
@@ -39,10 +39,11 @@ class AoclCrypto(CMakePackage):
 
     version(
-        "5.0",
-        sha256="b15e609943f9977e13f2d5839195bb7411c843839a09f0ad47f78f57e8821c23",
+        "5.1",
+        sha256="a2f768b7d37516c5c29cca0034aba90b91d02e477c762f2fa0fe4b1c30613973",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_crypto/package.py:
```diff
@@ -57,11 +58,11 @@ class AoclCrypto(CMakePackage):
     depends_on("c", type="build")  # generated
     depends_on("cxx", type="build")  # generated
-    depends_on("cmake@3.22:", type="build")
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_da/package.py:
```diff
@@ -57,9 +58,9 @@ class AoclDa(CMakePackage):
-
-    depends_on("cmake@3.22:", type="build")
-    for vers in ["5.0"]:
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_da/package.py:
```diff
@@ -34,6 +34,7 @@ class AoclDa(CMakePackage):
 
     maintainers("amd-toolchain-support")
 
+    version("5.1", sha256="93cdb789c948bf750e531f95618bae4370d53eddc88960744ff02c9acbfe9ef5")
```
Hurray ;-)

wdconinc commented on repos/spack_repo/builtin/packages/aocl_libmem/package.py:
```diff
@@ -35,10 +35,11 @@ class AoclLibmem(CMakePackage):
 
     version(
-        "5.0",
-        sha256="d3148db1a57fec4f3468332c775cade356e8133bf88385991964edd7534b7e22",
+        "5.1",
+        sha256="e03bc712a576b3e14ae433a696558e121dc67aac7fc1b4dca9b727605784e994",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_libmem/package.py:
```diff
@@ -47,22 +48,58 @@ class AoclLibmem(CMakePackage):
+    )
+    variant(
+        "vectorization",
+        default="none",
```
Note that for a while this would not actually give the correct default value since two identical variants for different when version ranges with different defaults deferred to the first default in the concretizer. @alalazo Is this still the case?

amd-toolchain-support replied:
I am not expert on this. 
However, during testing, I was able to get the right default value based on version, suggesting the variant is picked appropriately as shown in attached snippets:
- **_aocl-libmem@5.0_**
<img width="1798" height="417" alt="image" src="https://github.com/user-attachments/assets/e658a10e-8f5f-4814-83d2-01c03edbff00" />

- **_aocl-libmem@5.1_**
<img width="1807" height="391" alt="image" src="https://github.com/user-attachments/assets/4c6c897e-6dae-4825-9d74-3c002a8c7768" />

wdconinc commented on repos/spack_repo/builtin/packages/aocl_libmem/package.py:
```diff
@@ -47,22 +48,58 @@ class AoclLibmem(CMakePackage):
     depends_on("cxx", type="build")  # generated
 
-    depends_on("cmake@3.22:", type="build")
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_sparse/package.py:
```diff
@@ -32,10 +32,11 @@ class AoclSparse(CMakePackage):
 
     version(
-        "5.0",
-        sha256="7528970f41ae60563df9fe1f8cc74a435be1566c01868a603ab894e9956c3c94",
+        "5.1",
+        sha256="a5fff94f9144cb5e5e0f4702cc0d48a20215ccccd06cbed915e566e5d901fa0a",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_sparse/package.py:
```diff
@@ -73,7 +74,7 @@ class AoclSparse(CMakePackage):
     depends_on("boost", when="+benchmarks")
     depends_on("boost", when="@2.2")
-    depends_on("cmake@3.22:", type="build")
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_utils/package.py:
```diff
@@ -37,10 +37,11 @@ class AoclUtils(CMakePackage):
 
     version(
-        "5.0",
-        sha256="ee2e5d47f33a3f673b3b6fcb88a7ef1a28648f407485ad07b6e9bf1b86159c59",
+        "5.1",
+        sha256="68d75e04013abe90ea8308a9bc99b99532233b6c7f937f35381563f4124c20a5",
         preferred=True,
     )
```
Preferred doesn't do anything when it's on the most recent version.

amd-toolchain-support replied:
Updated as suggested

wdconinc commented on repos/spack_repo/builtin/packages/aocl_utils/package.py:
```diff
@@ -52,7 +53,7 @@ class AoclUtils(CMakePackage):
     depends_on("cxx", type="build")
 
-    depends_on("cmake@3.22:", type="build")
+    depends_on("cmake@3.26:", type="build")
```
Retain previous limit for older versions.

amd-toolchain-support replied:
Updated as suggested

wdconinc reviewed (commented):
I gave it a look after all. Seems all straightforward to me, and you're the maintainer so you're just looking for a review by someone who can approve this for merge.

A few comments (they are repeated in all packages... so apologies for the long list).

amd-toolchain-support committed:
preferred=True not required for latest versions

amd-toolchain-support committed:
Retain cmake version dependency for older versions

amd-toolchain-support committed:
Review comment: update amdlibm to allow cmake build_system for future versions

amd-toolchain-support committed:
Extend cmake version dependency for future versions

wdconinc reviewed (dismissed):
Looks good to me.

amd-toolchain-support committed:
style fix

amd-toolchain-support commented:
> I gave it a look after all. Seems all straightforward to me, and you're the maintainer so you're just looking for a review by someone who can approve this for merge.
> 
> A few comments (they are repeated in all packages... so apologies for the long list).

Thank you @wdconinc for your time and attention.
I have addressed the comments and pushed updates.
